File: src/weaponsprocurement/gui/StockReviewPendingTrade.java
package weaponsprocurement.gui;

final class StockReviewPendingTrade {
    private final String itemKey;
    private final String submarketId;
    private int quantity;

    private StockReviewPendingTrade(String itemKey, String submarketId, int quantity) {
        this.itemKey = itemKey;
        this.submarketId = submarketId;
        this.quantity = quantity;
    }

    static StockReviewPendingTrade create(String itemKey, String submarketId, int quantity) {
        return itemKey == null || itemKey.isEmpty() || quantity == 0
                ? null
                : new StockReviewPendingTrade(itemKey, submarketId, quantity);
    }

    String getItemKey() {
        return itemKey;
    }

    String getSubmarketId() {
        return submarketId;
    }

    int getQuantity() {
        return quantity;
    }

    void addQuantity(int amount) {
        quantity += amount;
    }

    boolean isBuy() {
        return quantity > 0;
    }

    boolean isSell() {
        return quantity < 0;
    }

    boolean isZero() {
        return quantity == 0;
    }

    boolean matches(String otherItemKey, String otherSubmarketId) {
        if (itemKey == null ? otherItemKey != null : !itemKey.equals(otherItemKey)) {
            return false;
        }
        return submarketId == null ? otherSubmarketId == null : submarketId.equals(otherSubmarketId);
    }

    StockReviewPendingTrade copy() {
        return create(itemKey, submarketId, quantity);
    }

    @Override
    public boolean equals(Object other) {
        if (this == other) {
            return true;
        }
        if (!(other instanceof StockReviewPendingTrade)) {
            return false;
        }
        StockReviewPendingTrade trade = (StockReviewPendingTrade) other;
        return quantity == trade.quantity && matches(trade.itemKey, trade.submarketId);
    }

    @Override
    public int hashCode() {
        int result = itemKey == null ? 0 : itemKey.hashCode();
        result = 31 * result + (submarketId == null ? 0 : submarketId.hashCode());
        result = 31 * result + quantity;
        return result;
    }
}

File: src/main/kotlin/weaponsprocurement/gui/StockReviewPendingTrade.kt
package weaponsprocurement.gui

class StockReviewPendingTrade private constructor(
    val itemKey: String,
    val submarketId: String?,
    var quantity: Int,
) {
    fun addQuantity(amount: Int) {
        quantity += amount
    }

    fun isBuy(): Boolean = quantity > 0

    fun isSell(): Boolean = quantity < 0

    fun isZero(): Boolean = quantity == 0

    fun matches(otherItemKey: String?, otherSubmarketId: String?): Boolean {
        if (itemKey != otherItemKey) return false
        return submarketId == otherSubmarketId
    }

    fun copy(): StockReviewPendingTrade? = create(itemKey, submarketId, quantity)

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is StockReviewPendingTrade) return false
        return quantity == other.quantity && matches(other.itemKey, other.submarketId)
    }

    override fun hashCode(): Int {
        var result = itemKey.hashCode()
        result = 31 * result + (submarketId?.hashCode() ?: 0)
        result = 31 * result + quantity
        return result
    }

    companion object {
        @JvmStatic
        fun create(itemKey: String?, submarketId: String?, quantity: Int): StockReviewPendingTrade? {
            return if (itemKey.isNullOrEmpty() || quantity == 0) {
                null
            } else {
                StockReviewPendingTrade(itemKey, submarketId, quantity)
            }
        }
    }
}
